File: Python/S_V_M_.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jul 13 13:58:39 2019

@author: RAMAN
"""

import os
import numpy as np
import pandas as pd

os.chdir("C:\\Users\\RAMAN\\Documents\\Python Scripts\\Data")
os.getcwd()

train_data = pd.read_csv('R_Module_Day_7.2_Credit_Risk_Train_data.csv')
test_data = pd.read_csv('R_Module_Day_8.2_Credit_Risk_Test_data.csv')

train_data['Source'] = 'Train'
test_data['Source'] = 'Test'
raw_data = pd.concat([train_data, test_data], axis = 0)
raw_data.shape

raw_data.dtypes
raw_data.isnull().sum()

raw_data['Dependents'].unique()
raw_data['Dependents'] = np.where(raw_data['Dependents'] == '3+', '3', raw_data['Dependents'])

# raw_data['Dependents'] = pd.factorize( [raw_data['Dependents']] )[0]

# raw_data['Dependents'].astype("category").cat.codes
# raw_data['Dependents'] = pd.Categorical()

raw_data['Dependents'] = pd.Series(raw_data['Dependents'], dtype = 'float64')

for i in raw_data.columns:
    if(raw_data[i].dtype == object):
        raw_data[i] = raw_data[i].fillna('Unknown')
        
    else:
        temp_imputation_value = raw_data.loc[raw_data['Source'] == "Train", i].median()
        raw_data[i].fillna(temp_imputation_value, inplace = True)

categ_vars = raw_data.loc[:, raw_data.dtypes == object].columns

dummy_df = pd.get_dummies(raw_data[categ_vars].drop(['Loan_ID', 'Source', 'Loan_Status'], axis=1), drop_first = True, dtype = int)
dummy_df.columns

full_data = pd.concat([raw_data, dummy_df], axis = 1)
Cols_To_Drop = categ_vars.drop(['Source','Loan_Status'])

full_data.shape
full_data2 = full_data.drop(Cols_To_Drop, axis = 1).copy()
full_data2.info()

description = full_data2.describe()

full_data2['Loan_Status']=np.where(full_data2['Loan_Status']== 'N',1,0)

fullraw3 = full_data2.copy()



var_list_for_detection = ['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount']

def outliers(df, var_list):
    for i in var_list:
        quant = i.quantile([0.25, 0.75])
        upper_bound = quant[0.75] + 1.5 * (quant[0.75] - quant[0.25])
        lower_bound = quant[0.25] - 1.5 * (quant[0.75] - quant[0.25])
        
        df[i] = np.where(df[i] > upper_bound, upper_bound, df[i])
        df[i] = np.where(df[i] < lower_bound, lower_bound, df[i])
    return (df)


def outlier_detection_correction(variable_list, data):
    for varName in variable_list:
        Q3 = np.percentile(data.loc[data['Source'] == 'Train', varName], 75)
        Q1 = np.percentile(data.loc[data['Source'] == 'Train', varName], 25)
        
        upper = Q3 + 1.5*(Q3-Q1)
        lower = Q1 - 1.5*(Q3-Q1)

        data[varName] = np.where(data[varName] > upper, upper, data[varName])
        data[varName] = np.where(data[varName] < lower, lower, data[varName])
    
    return (data)

before_summary = fullraw3[fullraw3['Source'] == 'Train'].describe()
Fullraw = outlier_detection_correction(var_list_for_detection, fullraw3)
#Fullraw = fullraw3.copy()
after_summary = Fullraw[fullraw3['Source'] == 'Train'].describe()

#####################################################
# Sampling - Train_X, Train_Y, Test_X, Test_Y
#####################################################

Train_X = Fullraw.loc[Fullraw['Source'] == 'Train'].drop(['Source', 'Loan_Status'], axis = 1).copy()

Train_Y = Fullraw.loc[Fullraw['Source'] == 'Train']
Train_Y = Train_Y['Loan_Status'].copy()
Train_X.shape

Test_X = Fullraw.loc[Fullraw['Source'] == 'Test'].drop(['Source', 'Loan_Status'], axis = 1).copy()

Test_Y = Fullraw.loc[Fullraw['Source'] == 'Test']
Test_Y = Test_Y['Loan_Status'].copy()
Test_X.shape

##########################################
# Model 2
##########################################

from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score

M2 = SVC() # Default kernel is 'rbf'

M2_Model = M2.fit(Train_X, Train_Y)
Test_class_M2 = M2_Model.predict(Test_X)
Confusion_Mat2 = confusion_matrix(Test_Y, Test_class_M2)
sum(np.diagonal(Confusion_Mat2))/Test_X.shape[0] * 100 # 79%

f1_score(Test_Y, Test_class_M2)

#########################
# Model 1 (kernel = linear)
# DO NOT TRY THIS AT HOME
#########################
 

#############################
# Manual Grid Searching
#############################

#############################
# grid Search - Alternate 1
#############################
df = {}
dataTable = pd.DataFrame()
cost = [1,2]
gamma = [0.01, 0.02]
kernel = ['sigmoid', 'rbf']

for krnl in kernel:
    for ct in cost:
        for gmma in gamma:
            M1 = SVC(kernel=krnl, C = ct, gamma = gmma)

            M1_Model = M1.fit(Train_X, Train_Y)
            Test_class_M1 = M1_Model.predict(Test_X)
            Confusion_Mat1 = confusion_matrix(Test_Y, Test_class_M1)
            accuracy = sum(np.diagonal(Confusion_Mat1))/Test_X.shape[0] * 100 
            df = pd.DataFrame({"Cost" : [ct],
                               "Kernel" : [krnl],
                               "Gamma" : [gmma],
                               "accuracy" : [accuracy]})
            dataTable = dataTable.append(df, ignore_index = True)
            

########################################
# python way of doing this( grid search )
########################################

from sklearn.model_selection import GridSearchCV

my_param_grid = {'C' : [1,2], 'gamma': [0.01, 0.1], 'kernel': ['sigmoid', 'rbf']} # it's a dictionary
SVM_GS = GridSearchCV(SVC(), param_grid= my_param_grid, scoring = 'accuracy', cv = 3)

SVM_GS_Model = SVM_GS.fit(Train_X, Train_Y)

# #Result in dictionary format
# SVM_GS_Model.cv_results_
# # Best tuning parameters
# SVM_GS_Model.best_params_

# Results in dataframe
SVM_Grid_Search_Df = pd.DataFrame.from_dict(SVM_GS_Model.cv_results_)

File: R/S_V_M_.R
install.packages("e1071")
library(DescTools)
library(e1071)
setwd("C:/Users/RAMAN/Documents/R")
TrainRaw = read.csv("R_Module_Day_7.2_Credit_Risk_Train_data.csv",na.strings = "") 

TestRaw = read.csv("R_Module_Day_8.2_Credit_Risk_Test_data.csv", na.strings = "")
TrainRaw$Source = "Train"
TestRaw$Source = "Test"

# Combine both datasets. Call it FullRaw
FullRaw = rbind(TrainRaw,TestRaw)

# Verify the results
dim(FullRaw)
colSums(is.na(FullRaw))
str(FullRaw)
levels(FullRaw$Dependents)

# Convert "Dependents" categorical indep var into a continuous variable

FullRaw$Dependents <- ifelse(FullRaw$Dependents == "3+", 3, c(0,1,2))
str(FullRaw)

# Perform missing value imputation for categorical variables by imputing the NAs with a new category called "Unknown"

# Gender
FullRaw$Gender <- factor(FullRaw$Gender, levels = c(levels(FullRaw$Gender), "Unknown"))
FullRaw$Gender[is.na(FullRaw$Gender)] = "Unknown"

# Married
FullRaw$Married <- as.character(FullRaw$Married)
FullRaw$Married[is.na(FullRaw$Married)] <- "Unknown"
FullRaw$Married <- as.factor(FullRaw$Married)

# Self-Employed
FullRaw$Self_Employed <- as.character(FullRaw$Self_Employed)
FullRaw$Self_Employed[is.na(FullRaw$Self_Employed)] <- "Unknown"
FullRaw$Self_Employed <- as.factor(FullRaw$Self_Employed)

str(FullRaw)
colSums(is.na(FullRaw))

# Perform missing value imputation for continuous variables
# Loan Amount
Mymedian <- median(FullRaw$LoanAmount[FullRaw$Source == "Train"], na.rm = TRUE)
FullRaw$LoanAmount[is.na(FullRaw$LoanAmount)] <- Mymedian

# Loan_Amount_Term
Mymedian <- median(FullRaw$Loan_Amount_Term[FullRaw$Source == "Train"], na.rm = TRUE)
FullRaw$Loan_Amount_Term[is.na(FullRaw$Loan_Amount_Term)] <- Mymedian

# Credit_History
Mymedian <- median(FullRaw$Credit_History[FullRaw$Source == "Train"], na.rm = TRUE)
FullRaw$Credit_History[is.na(FullRaw$Credit_History)] <- Mymedian

# Dependents
Mymedian <- median(FullRaw$Dependents[FullRaw$Source == "Train"], na.rm = TRUE)
FullRaw$Dependents[is.na(FullRaw$Dependents)] <- Mymedian
str(FullRaw)

Dummy_df = model.matrix(~Gender+Married+Education+Self_Employed+Property_Area, data = FullRaw)
View(Dummy_Df)
Dummy_Df = Dummy_df[,-c(1)]

FullRaw <- subset(FullRaw, select = -c(Gender, Married, Education, Self_Employed, Property_Area, Loan_ID))

Data = cbind(FullRaw, Dummy_Df)
 dim(Data)
 Data$Loan_Status = ifelse(Data$Loan_Status == "N", 1, 0)
str(Data)

TrainData <- subset(Data, Source == "Train")
TrainData <- subset(TrainData, select = -c(Source))

TestData <- subset(Data, Source == "Test")
TestData <- subset(TestData, select = -c(Source))
dim(TestData) ; dim(TrainData)

# SVM Modeling

# Model
M1 <- svm(as.factor(Loan_Status) ~ ., kernel = "linear", data = TrainData)

# Prediction on TestSet
M1_Test_Pred <- predict(M1, TestData)
vif(M1_Test_Pred)

# Confusion Matrix
table(M1_Test_Pred, TestData$Loan_Status)

# Tune svm on hyperparameters
obj <- tune.svm(as.factor(Loan_Status)~., data = TestData, gamma = c(0, 0.3, 0.6, 0.9), cost = c(1, 10, 100) )

# grid_search
kernel <- c("radial", "polynomial")
gamma <- c(0, 0.3, 0.6, 0.9, 1)
cost <- c(1, 10, 100)
mentalTable <- c()
for (i in kernel) {
  for (j in gamma) {
    for (k in cost) {
      M1 <- svm(as.factor(Loan_Status) ~ ., kernel = i, gamma = j, cost = k, data = TrainData)
      M1_Pred_Test <- predict(M1, TestData)
      confusion_matrix <- table(M1_Pred_Test, TestData$Loan_Status)
      accuracy <- sum(diag(confusion_matrix))/nrow(TestData)
      mentalTable <- rbind(mentalTable, c(kernel = i,gamma= j,cost=  k, accuracy =  accuracy))
    }
  }
}
View(mentalTable)
max(mentalTable$accuracy)
